File: day3-data-classes-sealed-classes-objects/java/CompanionObjectExample.java
/**
 * Day 3 - Companion Objects & Object Declarations in Java vs Kotlin
 *
 * Java:
 *   - Static methods and fields
 *   - Static utility classes
 *   - Singleton pattern (various implementations)
 *   - Factory methods
 *
 * Kotlin:
 *   - companion object — static-like members
 *   - object declaration — singleton
 *   - object expression — anonymous object
 *   - No "static" keyword!
 */
import java.util.Map;
import java.util.HashMap;

public class CompanionObjectExample {

    // ═══════════════════════════════════════════════════════════════
    // 1. JAVA: STATIC METHODS AND FIELDS
    // ═══════════════════════════════════════════════════════════════

    // Java: static keyword for class-level members
    public static class MathUtils {
        public static final double PI = 3.14159;
        public static final int MAX_SIZE = 100;

        // Static method
        public static int add(int a, int b) {
            return a + b;
        }

        // Static factory method
        public static MathUtils create() {
            return new MathUtils();
        }

        // Private constructor (prevent instantiation)
        private MathUtils() {}
    }

    // Java: Call static members via class name
    // MathUtils.add(1, 2)
    // MathUtils.PI

    // ═══════════════════════════════════════════════════════════════
    // 2. JAVA: SINGLETON (various implementations)
    // ═══════════════════════════════════════════════════════════════

    // Java: Eager initialization
    public static class Singletons {
        // Eager singleton
        public static final Singletons EAGER = new Singletons();

        // Lazy singleton (thread-safe)
        private static volatile Singletons instance;
        public static Singletons getInstance() {
            if (instance == null) {
                synchronized (Singletons.class) {
                    if (instance == null) {
                        instance = new Singletons();
                    }
                }
            }
            return instance;
        }

        // Enum singleton (best practice)
        public enum EnumSingleton {
            INSTANCE;
            public void doSomething() { System.out.println("Enum singleton"); }
        }

        // Holder singleton (lazy, thread-safe)
        private static class Holder {
            static final Singletons INSTANCE = new Singletons();
        }
        public static Singletons getHolderInstance() {
            return Holder.INSTANCE;
        }
    }

    // Java: Singleton is verbose and error-prone!

    // ═══════════════════════════════════════════════════════════════
    // 3. JAVA: FACTORY PATTERN
    // ═══════════════════════════════════════════════════════════════

    // Java: Factory method pattern
    public interface Shape {
        double area();
    }

    public static class Circle implements Shape {
        private final double radius;
        public Circle(double radius) { this.radius = radius; }
        public double area() { return Math.PI * radius * radius; }

        // Factory method
        public static Circle create(double radius) {
            return new Circle(radius);
        }
    }

    public static class Rectangle implements Shape {
        private final double width, height;
        public Rectangle(double width, double height) {
            this.width = width;
            this.height = height;
        }
        public double area() { return width * height; }

        // Factory method
        public static Rectangle create(double width, double height) {
            return new Rectangle(width, height);
        }
    }

    // ═══════════════════════════════════════════════════════════════
    // 4. COMPARISON TABLE
    // ═══════════════════════════════════════════════════════════════
    //
    //   Feature              Java                              Kotlin
    //   ──────────────────── ────────────────────────────────  ────────────────────────────
    //   Static members       static keyword                   companion object
    //   Static method        static void f()                  companion object { fun f() }
    //   Static field         static final int X = 5           companion object { val X = 5 }
    //   Singleton            Various patterns                 object : ClassName { }
    //   Factory              Static method                     companion object { fun create() }
    //   Type access          ClassName.staticMethod()         ClassName.staticMethod()
    //   Instance access      N/A                              ClassName().method()
    //
    // ═══════════════════════════════════════════════════════════════

    public static void main(String[] args) {

        // Java static members
        System.out.println("PI: " + MathUtils.PI);
        System.out.println("Add: " + MathUtils.add(1, 2));

        // Java singleton
        Singletons singleton = Singletons.getInstance();
        Singletons.EnumSingleton.INSTANCE.doSomething();

        // Java factory
        Shape circle = Circle.create(5.0);
        Shape rectangle = Rectangle.create(4.0, 6.0);
        System.out.println("Circle area: " + circle.area());
        System.out.println("Rectangle area: " + rectangle.area());

        // ═══════════════════════════════════════════════════════════════
        // KEY TAKEAWAYS
        // ═══════════════════════════════════════════════════════════════
        //
        // JAVA:
        //   - static keyword for class-level members
        //   - Singleton requires complex implementation
        //   - Factory pattern requires interface + implementations
        //   - No true "companion" concept
        //
        // KOTLIN:
        //   - companion object replaces static
        //   - object declaration = singleton (one line!)
        //   - Companion object can implement interfaces
        //   - Companion object can have factory methods
        //   - Much cleaner and more powerful
        //
        // ═══════════════════════════════════════════════════════════════
    }
}

File: day3-data-classes-sealed-classes-objects/kotlin/CompanionObjectExample.kt
/**
 * Day 3 - Companion Objects & Object Declarations in Kotlin
 *
 * Kotlin has NO "static" keyword! Instead:
 *   - companion object — static-like members
 *   - object declaration — singleton
 *   - object expression — anonymous object
 *
 * Benefits:
 *   - Companion object can implement interfaces
 *   - Can have factory methods
 *   - Can have companion object extensions
 *   - Much cleaner than Java's static!
 */

// ═══════════════════════════════════════════════════════════════
// 1. COMPANION OBJECT (replaces static)
// ═══════════════════════════════════════════════════════════════

// Kotlin: companion object replaces static members
class MathUtils private constructor() {

    // Companion object — like static members
    companion object {
        const val PI = 3.14159
        const val MAX_SIZE = 100

        // Factory method
        fun create(): MathUtils {
            return MathUtils()
        }

        // Static-like method
        fun add(a: Int, b: Int): Int = a + b
    }

    // Instance method
    fun calculate(): Double = PI * 2
}

// Java equivalent:
// public class MathUtils {
//     public static final double PI = 3.14159;
//     public static int add(int a, int b) { return a + b; }
//     public static MathUtils create() { return new MathUtils(); }
// }

// ═══════════════════════════════════════════════════════════════
// 2. COMPANION OBJECT WITH INTERFACE
// ═══════════════════════════════════════════════════════════════

// Kotlin: Companion object can implement interfaces!
interface Logger {
    fun log(message: String)
}

class AppConfig private constructor() {
    val host: String = "localhost"
    val port: Int = 8080

    // Companion object can implement interface
    companion object : Logger {
        const val VERSION = "1.0"

        override fun log(message: String) {
            println("[AppConfig] $message")
        }

        // Factory method
        fun create(): AppConfig {
            log("Creating AppConfig")
            return AppConfig()
        }
    }
}

// Java: Would need static utility class + separate logger

// ═══════════════════════════════════════════════════════════════
// 3. COMPANION OBJECT EXTENSIONS
// ═══════════════════════════════════════════════════════════════

// Kotlin: Can add extension functions to companion object
class User(val name: String, val age: Int) {
    companion object {
        // Factory methods
        fun fromName(name: String) = User(name, 0)
        fun create(name: String, age: Int) = User(name, age)
    }
}

// Extension function on companion object
fun User.Companion.guest(): User = User("Guest", 0)
fun User.Companion.fromEmail(email: String): User {
    val name = email.substringBefore("@")
    return User(name, 0)
}

// Java: Cannot extend static members!

// ═══════════════════════════════════════════════════════════════
// 4. OBJECT DECLARATION (Singleton)
// ═══════════════════════════════════════════════════════════════

// Kotlin: object declaration = thread-safe singleton (one line!)
object DatabaseConfig {
    val host = "localhost"
    val port = 5432
    val maxConnections = 10

    fun connect(): String {
        return "Connecting to $host:$port"
    }
}

// Java equivalent:
// public class DatabaseConfig {
//     private static volatile DatabaseConfig instance;
//     public static DatabaseConfig getInstance() {
//         if (instance == null) {
//             synchronized (DatabaseConfig.class) {
//                 if (instance == null) {
//                     instance = new DatabaseConfig();
//                 }
//             }
//         }
//         return instance;
//     }
// }

// ═══════════════════════════════════════════════════════════════
// 5. OBJECT EXPRESSION (Anonymous Object)
// ═══════════════════════════════════════════════════════════════

// Kotlin: object expression = anonymous object (like Java's anonymous class)
interface ClickListener {
    fun onClick(viewId: String)
}

fun setupButton() {
    // Kotlin: object expression
    val listener = object : ClickListener {
        override fun onClick(viewId: String) {
            println("Clicked: $viewId")
        }
    }

    // Java equivalent:
    // ClickListener listener = new ClickListener() {
    //     @Override
    //     public void onClick(String viewId) {
    //         System.out.println("Clicked: " + viewId);
    //     }
    // };
}

// Kotlin: Can create anonymous object without interface
fun createAnonymousObject() {
    val obj = object {
        val x = 10
        val y = 20
        fun sum() = x + y
    }
    println("Sum: ${obj.sum()}")

    // Java: Cannot create anonymous object without interface/class!
    // Must use some class or interface
}

// ═══════════════════════════════════════════════════════════════
// 6. PRACTICAL EXAMPLES
// ═══════════════════════════════════════════════════════════════

// Example 1: Factory Pattern
sealed class Shape {
    data class Circle(val radius: Double) : Shape()
    data class Rectangle(val width: Double, val height: Double) : Shape()

    companion object {
        // Factory methods
        fun circle(radius: Double) = Circle(radius)
        fun rectangle(width: Double, height: Double) = Rectangle(width, height)

        // From string
        fun fromString(description: String): Shape {
            val parts = description.split(",")
            return when (parts[0]) {
                "circle" -> Circle(parts[1].toDouble())
                "rectangle" -> Rectangle(parts[1].toDouble(), parts[2].toDouble())
                else -> throw IllegalArgumentException("Unknown shape: $description")
            }
        }
    }
}

// Example 2: Constants Holder
object Constants {
    const val APP_NAME = "MyApp"
    const val VERSION = "1.0.0"
    const val API_TIMEOUT = 30000L

    object Paths {
        const val API_BASE = "/api/v1"
        const val USERS = "$API_BASE/users"
        const val POSTS = "$API_BASE/posts"
    }
}

// Example 3: Service Locator
object ServiceLocator {
    private val services = mutableMapOf<String, Any>()

    fun <T : Any> register(name: String, service: T) {
        services[name] = service
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> get(name: String): T = services[name] as T
}

// ═══════════════════════════════════════════════════════════════
// MAIN FUNCTION
// ═══════════════════════════════════════════════════════════════

fun main() {

    // Companion object
    println("PI: ${MathUtils.PI}")
    println("Add: ${MathUtils.add(1, 2)}")
    val utils = MathUtils.create()
    println("Calculate: ${utils.calculate()}")

    // Companion object with interface
    val config = AppConfig.create()
    println("Host: ${config.host}")
    println("Version: ${AppConfig.VERSION}")
    AppConfig.log("App started")

    // Companion object extensions
    val guest = User.guest()
    val fromEmail = User.fromEmail("alice@example.com")
    println("Guest: ${guest.name}")
    println("From email: ${fromEmail.name}")

    // Object declaration (singleton)
    println("Connect: ${DatabaseConfig.connect()}")
    println("Max connections: ${DatabaseConfig.maxConnections}")

    // Object expression
    setupButton()
    createAnonymousObject()

    // Factory pattern
    val circle = Shape.circle(5.0)
    val rect = Shape.rectangle(4.0, 6.0)
    val fromStr = Shape.fromString("circle,3.0")
    println("Circle: $circle")
    println("Rectangle: $rect")
    println("From string: $fromStr")

    // Constants
    println("App: ${Constants.APP_NAME} v${Constants.VERSION}")
    println("Users URL: ${Constants.Paths.USERS}")

    // Service locator
    ServiceLocator.register("database", "PostgreSQL")
    val db: String = ServiceLocator.get("database")
    println("Database: $db")

    // ═══════════════════════════════════════════════════════════════
    // COMPARISON SUMMARY
    // ═══════════════════════════════════════════════════════════════
    //
    // JAVA                              KOTLIN
    // ────────────────────────────────  ──────────────────────────────
    // static final int X = 5            companion object { const val X = 5 }
    // static void f() { }               companion object { fun f() { } }
    // Singleton (complex)               object Singleton { }
    // Anonymous class                   object : Interface { }
    // ClassName.staticMethod()          ClassName.staticMethod()
    // Math.random()                     Math.random() or Utils.random()
    //
    // KEY RULES:
    // ─────────────────────────────────────────────────
    // 1. companion object replaces static
    // 2. object declaration = singleton
    // 3. object expression = anonymous object
    // 4. Companion object can implement interfaces
    // 5. Can add extensions to companion object
    // 6. const val in companion object = static final
    // 7. @JvmStatic for Java interop
    // 8. @JvmField for direct field access
    //
    // ═══════════════════════════════════════════════════════════════
}
